dev/src/bulk-writer.ts

```diff
@@ -73,124 +76,91 @@ const RATE_LIMITER_MULTIPLIER = 1.5;
  */
 const RATE_LIMITER_MULTIPLIER_MILLIS = 5 * 60 * 1000;
 
-/*!
- * Used to represent the state of batch.
- *
- * Writes can only be added while the batch is OPEN. For a batch to be sent,
- * the batch must be READY_TO_SEND. After a batch is sent, it is marked as SENT.
- */
-enum BatchState {
-  OPEN,
-  READY_TO_SEND,
-  SENT,
-}
-
 /**
  * Used to represent a batch on the BatchQueue.
  *
  * @private
  */
-class BulkCommitBatch {
-  /**
-   * The state of the batch.
-   */
-  state = BatchState.OPEN;
-
+class BulkCommitBatch extends WriteBatch {
+  constructor(firestore: Firestore, readonly maxBatchSize: number) {
+    super(firestore);
+  }
   // The set of document reference paths present in the WriteBatch.
   readonly docPaths = new Set<string>();
 
   // An array of pending write operations. Only contains writes that have not
   // been resolved.
   private pendingOps: Array<Deferred<BatchWriteResult>> = [];
 
-  private readonly backoff: ExponentialBackoff;
-
-  constructor(
-    private readonly firestore: Firestore,
-    private writeBatch: WriteBatch,
-    private readonly maxBatchSize: number
-  ) {
-    this.backoff = new ExponentialBackoff();
+  _has(documentRef: firestore.DocumentReference): boolean {
+    return this.docPaths.has(documentRef.path);
   }
 
-  /**
-   * The number of writes in this batch.
-   */
-  get opCount(): number {
-    return this.pendingOps.length;
+  _markReadyToSend(): void {
+    if (this.state === BatchState.OPEN) {
```

Comment:
We should try to unify this: Either `state` lives in BulkCommitBatch and the helpers are here, or we should move the helpers to WriteBatch as well. What do you think?

Reply:
It's a little messy right now, since the WriteBatch uses the `BatchState.SENT` after the batch has been committed, but does not adhere to the READY_TO_SEND checks.

We have two options:
1. Move all the BatchState code into BulkCommitBatch and add back a `committed` boolean flag in WriteBatch. (more code, appropriate separation)
2. Move all the BatchState over into WriteBatch. (slightly less code, less-intuitive behavior)

Reply:
I opted to move everything back into BulkCommitBatch, which only results in an extra `_commited` variable on WriteBatch. The main issue I had with option 2 is that the helpers that I moved over are only used by BulkWriter, which didn't seem right.